File: html5/js/party_prop.js
var PartyProp = JunctionProps.Prop.extend(
	{
		init: function(propName){
			var self = this;
			this._super(propName, new this.PartyPropState(null),
						propName + "_" + randomUUID());

			this.topVideoChangedListeners = [];
			this.currentTopVideo = null;

			this.addChangeListener({ type: "change",
									 onChange: function(o){
										 self.updateOnChange();
									 }});
			this.addChangeListener({ type: "sync",
									 onChange: function(o){
										 self.updateOnChange();
									 }});
		},
		
		/** 
		 * Notify app of playlist change.
		 */
		updateOnChange: function(){
			var pl = this.getPlaylist();
			if(pl.length > 0){
				var top = pl[0];
				if(this.currentTopVideo == null || top.id != this.currentTopVideo.id){
					this.currentTopVideo = top;
					for(var i = 0; i < this.topVideoChangedListeners.length; i++){
						(this.topVideoChangedListeners[i])(this.currentTopVideo);
					}
				}
			}
			else {
				this.currentTopVideo = null;
			}
		},

		/** 
		 * Register a listener for this app-level event.
		 */
		addTopVideoChangedListener: function(func){
			this.topVideoChangedListeners.push(func);
		},


		reifyState: function(jsonObj){
			return new this.PartyPropState(jsonObj);
		},

		getName: function(){
			return this.state.getName();
		},

		setName: function(name){
			this.addOperation({type:"setName", name: name});
		},

		addObject: function(item){
			this.addOperation({type:"addObj", item:item});
		},

		deleteObject: function(item){
			this.addOperation({type:"deleteObj", itemId: item.id});
		},

		addPicture: function(userId,url,thumbUrl,caption){
			this.addObject(
				{
					id: randomUUID(),
					type: "image",
					url: url,
					thumbUrl: thumbUrl,
					time: Math.ceil((new Date()).getTime() / 1000.0),
					caption: caption,
					owner: userId
				});
		},

		recycleTopVideo: function(){ 
			if(this.currentTopVideo != null){
				var v = this.currentTopVideo;
				this.addYoutube(v.id, v.owner, v.videoId, v.thumbUrl, v.caption);
			}
		},

		addYoutube: function(id, userId,videoId,thumbUrl,caption){
			this.addObject(
				{
					id: id || randomUUID(),
					type: "youtube",
					videoId: videoId,
					thumbUrl: thumbUrl,
					time: Math.ceil((new Date()).getTime() / 1000),
					caption: caption,
					owner: userId,
					votes: 0
				});
		},

		eachObject: function(iterator){
			this.state.eachObject(iterator);
		},
		getPictures: function(){
			return this.state.getPictures();
		},
		getUsers: function(){
			return this.state.getUsers();
		},
		getPlaylist: function(){
			return this.state.getPlaylist();
		},

		PartyPropState: Class.extend(
			{
				init: function(jsonObj){
					if(jsonObj == null){
						this.raw = {
							name: "Unnamed Party",
							objects: {}
						};
					}
					else{
						this.raw = jsonObj;	
					}
				},
				applyOperation: function(op){
					if(op.type == "addObj"){
						var obj = op.item;
						this.raw.objects[obj.id] = obj;
					}
					else if(op.type == "deleteObj"){
						var id = op.itemId;
						delete this.raw.objects[id];
					}
					else if(op.type == "setName"){
						this.raw.name = op.name;
					}
					else if(op.type == "vote"){
						var count = op.count;
						var obj = this.raw.objects[op.itemId];
						if(obj){
							var cur = obj.votes || 0;
							obj.votes = cur + count;
						}
					}
				},

				eachObject: function(iterator){
					for(var id in this.raw.objects){
						var obj = this.raw.objects[id];
						iterator(obj);
					}
				},

				getName: function(){ return this.raw.name; },

				objectsOfType: function(tpe){
					var objs = [];
					for(var id in this.raw.objects){
						var obj = this.raw.objects[id];
						if(obj.type == tpe){
							objs.push(obj);
						}
					}
					return objs;
				},

				getPictures: function(){
					var pics = this.objectsOfType("image");
					pics.sort(function(a,b){ return b.time - a.time; });
					return pics;
				},

				getUsers: function(){
					var users = this.objectsOfType("user");
					return users;
				},

				getPlaylist: function(){
					var vids = this.objectsOfType("youtube");
					vids.sort(function(a,b){ 
								  if(a.votes == b.votes) return (a.time - b.time);
								  else return b.votes - a.votes;
							  });
					return vids;
				},

				toJSON:function(){
					return JSON.parse(JSON.stringify(this.raw));
				},

				hashCode: function(){
					return 1;
				},

				copy: function(){
					return new PartyProp.prototype.PartyPropState(this.toJSON());
				}
			})

	});

File: android/src/edu/stanford/junction/sample/partyware/PartyProp.java
package edu.stanford.junction.sample.partyware;

import java.util.*;
import org.json.JSONObject;
import org.json.JSONException;
import edu.stanford.junction.props2.*;
import edu.stanford.junction.extra.JSONObjWrapper;

public class PartyProp extends Prop {

	public PartyProp(String propName, String propReplicaName, IPropState s){
		super(propName, propReplicaName, s);
	}

	public PartyProp(String propName){
		this(propName, propName + (new Random()).nextInt(), new PartyState());
	}

	public IProp newFresh(){
		return new PartyProp(getPropName(), getPropReplicaName(), new PartyState());
	}

	public void forceChangeEvent(){
		dispatchChangeNotification(EVT_SYNC, null);
	}

	protected IPropState reifyState(JSONObject obj){
		return new PartyState(obj);
	}

	protected JSONObject newAddObjOp(JSONObject item){
		JSONObject obj = new JSONObject();
		try{
			obj.put("type", "addObj");
			obj.put("item", item);
		}catch(JSONException e){}
		return obj;
	}

	protected JSONObject newVoteOp(String itemId, int count){
		JSONObject obj = new JSONObject();
		try{
			obj.put("type", "vote");
			obj.put("itemId", itemId);
			obj.put("count", count);
		}catch(JSONException e){}
		return obj;
	}

	protected void addObj(JSONObject item){
		addOperation(newAddObjOp(item));
	}


	/////////////////////////
    // Conveniance helpers //
    /////////////////////////

	public String getName(){
		return ((PartyState)getState()).getName();
	}

	public List<JSONObject> getImages(){
		PartyState s = (PartyState)getState();
		return s.getImages();
	}

	public List<JSONObject> getYoutubeVids(){
		PartyState s = (PartyState)getState();
		return s.getYoutubeVids();
	}

	public void addImage(String userId, String url, String thumbUrl, String caption, long time){
		addObj(newImageObj(userId, url, thumbUrl, caption, time));
	}

	public void upvoteVideo(String id){
		addOperation(newVoteOp(id, 1));
	}

	public void downvoteVideo(String id){
		addOperation(newVoteOp(id, -1));
	}

	public void addYoutube(String userId, String videoId, String thumbUrl, String caption, long time){
		addObj(newYoutubeObj(userId, videoId, thumbUrl, caption, time));
	}

	protected JSONObject newImageObj(String userId, String url, String thumbUrl, String caption, long time){
		JSONObject obj = newHTTPResourceObj("image", userId, url, caption, time);
		try{
			obj.put("thumbUrl", thumbUrl);
		}catch(JSONException e){}
		return obj;
	}

	protected JSONObject newYoutubeObj(String userId, String videoId, 
									   String thumbUrl, String caption, long time){
		JSONObject obj = newHTTPResourceObj("youtube", userId, null, caption, time);
		try{
			obj.put("videoId", videoId);
			obj.put("thumbUrl", thumbUrl);
		}catch(JSONException e){}
		return obj;
	}

	protected JSONObject newHTTPResourceObj(String type, String userId, 
											String url, String caption, long time){
		JSONObject obj = new JSONObject();
		try{
			obj.put("id", (UUID.randomUUID()).toString());
			obj.put("type", type);
			obj.put("url", url);
			obj.put("time", time);
			obj.put("caption", caption);
			obj.put("owner", userId);
		}catch(JSONException e){}
		return obj;
	}


	//////////////////////////
    // The State definition //
    //////////////////////////

	static class PartyState implements IPropState {

		private HashMap<String, JSONObject> objects = new HashMap<String, JSONObject>();
		private String name = "Unnamed Party";

		public PartyState(PartyState other){
			if(other != null){
				this.name = other.getName();
				Iterator it = other.objects.entrySet().iterator();
				while (it.hasNext()) {
					Map.Entry<String,JSONObject> pair = 
						(Map.Entry<String,JSONObject>)it.next();
					if(pair.getValue() instanceof JSONObjWrapper){
						JSONObjWrapper obj = (JSONObjWrapper)pair.getValue();
						objects.put(pair.getKey(), (JSONObject)obj.clone());
					}
				}
			}
		}

		public PartyState(JSONObject obj){
			this.name = obj.optString("name");
			JSONObject jsonObjects = obj.optJSONObject("objects");
			if(jsonObjects != null){
				Iterator it = jsonObjects.keys();
				while(it.hasNext()){
					Object key = it.next();
					Object val = jsonObjects.opt(key.toString());
					if(val instanceof JSONObject){
						JSONObject eaObj = (JSONObject)val;
						String id = eaObj.optString("id");
						objects.put(id, new JSONObjWrapper(eaObj));
					}
				}
			}
		}

		public PartyState(){
			this((PartyState)null);
		}

		public String getName(){
			return this.name;
		}
		
		public IPropState applyOperation(JSONObject op){
			String type = op.optString("type");
			if(type.equals("addObj")){
				JSONObject item = op.optJSONObject("item");
				String id = item.optString("id");
				objects.put(id, (new JSONObjWrapper(item)));
			}
			else if(type.equals("deleteObj")){
				String id = op.optString("itemId");
				objects.remove(id);
			}
			else if(type.equals("setName")){
				String name = op.optString("name");
				this.name = name;
			}
			else if(type.equals("vote")){
				String id = op.optString("itemId");
				int count = op.optInt("count");
				JSONObject o = objects.get(id);
				if(o != null){
					int cur = o.optInt("votes");
					System.out.println("CURRENT VOTES: " + cur);
					try{
						o.put("votes", cur + count);
					}
					catch(JSONException e){
						e.printStackTrace(System.err);
					}
					System.out.println("NEW VOTES: " + o.optInt("votes"));
				}
				else{
					System.err.println("Couldn't find object for id: " + id);
				}
			}
			else{
				throw new IllegalStateException("Unrecognized operation: " + type);
			}
			return this;
		}

		public JSONObject toJSON(){
			JSONObject obj = new JSONObject();
			JSONObject jsonObjects = new JSONObject();
			try{
				obj.put("name", this.name);
				obj.put("objects", jsonObjects);
			}
			catch(JSONException e){
				e.printStackTrace(System.err);
				throw new IllegalStateException("toJson failed in PartyProp!");
			}
			Iterator it = objects.entrySet().iterator();
			while (it.hasNext()) {
				Map.Entry<String,JSONObject> pair = 
					(Map.Entry<String,JSONObject>)it.next();
				try{
					JSONObjWrapper wrapper = (JSONObjWrapper)pair.getValue();
					jsonObjects.put(String.valueOf(pair.getKey()), 
									wrapper.getRaw());
				}
				catch(JSONException e){
					e.printStackTrace(System.err);
					throw new IllegalStateException("toJson failed in PartyProp!");
				}
			}
			return obj;
		}

		public List<JSONObject> getImages(){
			ArrayList<JSONObject> images = new ArrayList<JSONObject>();
			Iterator<JSONObject> it = objects.values().iterator();
			while (it.hasNext()) {
				JSONObject ea = it.next();
				String type = ea.optString("type");
				if(type.equals("image")){
					images.add(ea);
				}
			}
			sortByTime(images, true);
			return Collections.unmodifiableList(images);
		}

		private void sortByTime(List<JSONObject> input, final boolean newToOld){
			Collections.sort(input, new Comparator<JSONObject>(){
					public int compare(JSONObject o1, JSONObject o2) {
						if(newToOld){
							return (int)(o2.optLong("time") - o1.optLong("time"));
						}
						else{
							return (int)(o1.optLong("time") - o2.optLong("time"));
						}
					}
				});
		}

		private void sortByVotes(List<JSONObject> input, final boolean highToLow){
			Collections.sort(input, new Comparator<JSONObject>(){
					public int compare(JSONObject o1, JSONObject o2) {
						if(highToLow){
							return (int)(o2.optInt("votes") - o1.optInt("votes"));
						}
						else{
							return (int)(o1.optInt("votes") - o2.optInt("votes"));
						}
					}
				});
		}


		public List<JSONObject> getYoutubeVids(){
			ArrayList<JSONObject> vids = new ArrayList<JSONObject>();
			Iterator<JSONObject> it = objects.values().iterator();
			while (it.hasNext()) {
				JSONObject ea = it.next();
				String type = ea.optString("type");
				if(type.equals("youtube")){
					vids.add(ea);
				}
			}
			sortByVotes(vids, true);
			return Collections.unmodifiableList(vids);
		}

		public IPropState copy(){
			return new PartyState(this);
		}
	}

}